bundles/org.openhab.binding.tapocontrol/src/main/java/org/openhab/binding/tapocontrol/internal/TapoDiscoveryService.java

```diff
@@ -0,0 +1,265 @@
+import org.openhab.core.thing.ThingTypeUID;
+import org.openhab.core.thing.ThingUID;
+import org.openhab.core.thing.binding.ThingHandler;
+import org.openhab.core.thing.binding.ThingHandlerService;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.gson.JsonArray;
+import com.google.gson.JsonElement;
+import com.google.gson.JsonObject;
+
+/**
+ * Handler class for TAPO Smart Home thing discovery
+ *
+ * @author Christian Wild - Initial contribution
+ */
+@NonNullByDefault
+public class TapoDiscoveryService extends AbstractDiscoveryService implements ThingHandlerService {
+    private final Logger logger = LoggerFactory.getLogger(TapoDiscoveryService.class);
+    protected @Nullable ScheduledFuture<?> scanJob;
+    protected @Nullable TapoBridgeHandler bridge;
+
+    /***********************************
+     *
+     * INITIALIZATION
+     *
+     ************************************/
+
+    /**
+     * INIT CLASS
+     * 
+     * @param bridgeHandler
+     */
+    public TapoDiscoveryService() {
+        super(SUPPORTED_THING_TYPES_UIDS, TAPO_DISCOVERY_TIMEOUT_S, false);
+    }
+
+    /**
+     * activate
+     */
+    public void activate() {
+        activate(new HashMap<>());
+    }
+
+    /**
+     * deactivate
+     */
+    @Override
+    public void deactivate() {
+        stopScheduler(this.scanJob);
+        super.deactivate();
+    }
+
+    @Override
+    public void setThingHandler(@Nullable ThingHandler handler) {
+        if (handler instanceof TapoBridgeHandler) {
+            this.bridge = (TapoBridgeHandler) handler;
+        }
+    }
+
+    @Override
+    public @Nullable ThingHandler getThingHandler() {
+        return this.bridge;
+    }
+
+    /***********************************
+     *
+     * SCAN HANDLING
+     *
+     ************************************/
+
+    /**
+     * Start scan manually
+     */
+    @Override
+    public void startScan() {
+        removeOlderResults(getTimestampOfLastScan());
+        try {
+            JsonArray jsonArray = bridge.getDeviceList();
+            handleCloudDevices(jsonArray);
+        } catch (Exception e) {
+            logger.debug("Error scanning for devices", e);
+        }
+    }
+
+    /**
+     * Stop Scan
+     */
+    @Override
+    protected synchronized void stopScan() {
+        super.stopScan();
+    }
+
+    /**
+     * Abort Scan
+     */
+    @Override
+    public synchronized void abortScan() {
+        super.abortScan();
+    }
+
+    /**
+     * SCAN ONCE
+     * scan an disable scheduler
+     */
+    protected void startScanOnce() {
+        this.startScan();
+        stopScheduler(this.scanJob);
+    }
+
+    /**
+     * Stop scheduler
+     * 
+     * @param scheduler ScheduledFeature<?> which schould be stopped
+     */
+    protected void stopScheduler(@Nullable ScheduledFuture<?> scheduler) {
+        if (scheduler != null) {
+            scheduler.cancel(true);
+            scheduler = null;
```

Comment:
This assignment has no effect as it is a local variable.

Reply:
will be deleted in the next push-request